controllers/subscription_controller.go

```diff
@@ -18,45 +18,378 @@ package controllers
+	}
+	newSub := sub.DeepCopy()
+	now := metav1.Now()
+	newSub.Status.RepositoryHealth = corev1alpha1.RepositoryHealth{
+		RepositoryRef: &corev1.ObjectReference{Name: repo.Name, Namespace: repo.Namespace},
+		LastUpdated:   &now,
+		Healthy:       &healthy,
+	}
+	newSub.Status.SetConditions(corev1alpha1.SubscriptionReconcileSuccess(corev1alpha1.SubscriptionTypeSourceSynced))
+	if err = r.Status().Patch(ctx, newSub, client.MergeFrom(sub)); err != nil {
+		logger.Error(err, "Failed to patch subscription status repositoryHealth")
+		return err
+	}
+
+	return nil
+}
+
+// CreateComponentPlan create component plan if not exists or update component plan if exists
+func (r *SubscriptionReconciler) CreateComponentPlan(ctx context.Context, sub *corev1alpha1.Subscription, fetch corev1alpha1.ComponentVersion) error {
+	plan := &corev1alpha1.ComponentPlan{}
+	plan.Name = corev1alpha1.GenerateComponentPlanName(sub, fetch.Version)
+	plan.Namespace = sub.Namespace
+	plan.Spec.Override = sub.Spec.Override
+	plan.Spec.ComponentRef = sub.Spec.ComponentRef
+	plan.Spec.InstallVersion = fetch.Version
+	if sub.Spec.ComponentPlanInstallMethod.IsAuto() {
+		plan.Spec.Approved = true
+	}
+	return r.Create(ctx, plan)
+}
+
+// UpdateStatusInstalled update subscription status installed
+func (r *SubscriptionReconciler) UpdateStatusInstalled(ctx context.Context, logger logr.Logger, sub *corev1alpha1.Subscription, fetch corev1alpha1.ComponentVersion) (err error) {
+	plan := &corev1alpha1.ComponentPlan{}
+	planName := corev1alpha1.GenerateComponentPlanName(sub, fetch.Version)
+	planNs := sub.Namespace
+	if err = r.Get(ctx, types.NamespacedName{Namespace: planNs, Name: planName}, plan); err != nil {
+		logger.Error(err, "Failed to get componentPlan", "ComponentPlan.Namespace", planNs, "ComponentPlan.Name", planName)
+		if err1 := r.PatchCondition(ctx, sub, corev1alpha1.SubscriptionReconcileError(corev1alpha1.SubscriptionTypePlanSynce, err)); err1 != nil {
+			return errors.Wrap(err, err1.Error())
+		}
+		return err
+	}
+	newSub := sub.DeepCopy()
+	// latest should be first item
+	newSub.Status.Installed = append([]corev1alpha1.Installed{
+		{
+			InstalledVersion: fetch,
+			InstalledTime:    metav1.Now(),
+			ComponentPlanRef: &corev1.ObjectReference{Name: plan.Name, Namespace: plan.Namespace},
+		},
+	}, sub.Status.Installed...)
+	newSub.Status.SetConditions(corev1alpha1.SubscriptionReconcileSuccess(corev1alpha1.SubscriptionTypePlanSynce))
+	return r.Status().Patch(ctx, newSub, client.MergeFrom(sub))
+}
+
+// PatchCondition patch subscription status condition
+func (r *SubscriptionReconciler) PatchCondition(ctx context.Context, sub *corev1alpha1.Subscription, condition corev1alpha1.Condition) (err error) {
+	newSub := sub.DeepCopy()
+	newSub.Status.SetConditions(condition)
+	ready := true
+	for _, cond := range newSub.Status.Conditions {
+		if cond.Type == corev1alpha1.TypeReady {
+			continue
+		}
+		if cond.Status != corev1.ConditionTrue {
+			ready = false
```

Comment:
If ready is false,  maybe the loop should just stop, there is no other case to change ready to true.

Reply:
There are 3 condition type: 
- Ready (final one)
- SourceSynced (indicates that the subscription is synced from component and repository )
- PlanSynced (indicates that the componentplan is synced from subscription)

I think "Ready" is the final result, and users and other controllers can use this condition type to determine whether the subscription is working properly.
SourceSynced and PlanSynced only explain in detail whether the internal two stages are normal.

Reply:
I understand what you mean. What I mean is, add a break statement. There is no other case to set ready to true. It is possible to end the for loop early